SameServerEntity ports not mapped

Try to deploy this app to a Docker container:

``` yaml
name: App
location: my-docker-cloud
services:
- type: brooklyn.entity.basic.SameServerEntity
  name: Redises
  brooklyn.children:
  - type: brooklyn.entity.nosql.redis.RedisStore
```

I observe this NullPointerException:

```
2014-06-19 12:29:23,031 DEBUG b.location.jclouds.JcloudsUtil [brooklyn-execmanager-LPBGewXV-121]: Docker will forward these ports {22/tcp=[{HostIp=0.0.0.0, HostPort=49156}]}
2014-06-19 12:29:23,031 DEBUG b.l.d.DockerContainerLocation [brooklyn-execmanager-LPBGewXV-121]: Docker mapped port 6379 to null for Container 757d1a8a5c7d40cd3397fdfb6ba074d6dcf6c7ebe5d5efc9c8df06db2031364b
2014-06-19 12:29:23,032 DEBUG b.u.task.BasicExecutionManager [brooklyn-execmanager-LPBGewXV-121]: Exception running task Task[pre-start [SUB-TASK, Wrapped[contextEntity:RedisStoreImpl{id=l6T3DxDI}]]; e8wKFo44] (rethrowing): Cannot convert config value 6379-65535 for sensor Sensor: redis.port (java.lang.Integer): java.lang.NullPointerException
java.lang.IllegalArgumentException: Cannot convert config value 6379-65535 for sensor Sensor: redis.port (java.lang.Integer): java.lang.NullPointerException
    at brooklyn.event.basic.AttributeSensorAndConfigKey.getAsSensorValue(AttributeSensorAndConfigKey.java:95) ~[brooklyn-core-0.7.0-SNAPSHOT.jar:na]
    at brooklyn.event.feed.ConfigToAttributes.apply(ConfigToAttributes.java:28) ~[brooklyn-core-0.7.0-SNAPSHOT.jar:na]
    at brooklyn.event.feed.ConfigToAttributes.apply(ConfigToAttributes.java:17) ~[brooklyn-core-0.7.0-SNAPSHOT.jar:na]
    at brooklyn.entity.software.MachineLifecycleEffectorTasks.preStartCustom(MachineLifecycleEffectorTasks.java:349) ~[brooklyn-software-base-0.7.0-SNAPSHOT.jar:na]
    at brooklyn.entity.basic.SoftwareProcessDriverLifecycleEffectorTasks.preStartCustom(SoftwareProcessDriverLifecycleEffectorTasks.java:69) ~[brooklyn-software-base-0.7.0-SNAPSHOT.jar:na]
    at brooklyn.entity.software.MachineLifecycleEffectorTasks$6.run(MachineLifecycleEffectorTasks.java:289) ~[brooklyn-software-base-0.7.0-SNAPSHOT.jar:na]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) ~[na:1.7.0_51]
    at brooklyn.util.task.DynamicSequentialTask$DstJob.call(DynamicSequentialTask.java:296) ~[brooklyn-core-0.7.0-SNAPSHOT.jar:na]
    at brooklyn.util.task.BasicExecutionManager$3.call(BasicExecutionManager.java:353) ~[brooklyn-core-0.7.0-SNAPSHOT.jar:na]
    at java.util.concurrent.FutureTask.run(FutureTask.java:262) [na:1.7.0_51]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_51]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_51]
    at java.lang.Thread.run(Thread.java:744) [na:1.7.0_51]
Caused by: java.lang.NullPointerException: null
    at brooklyn.location.docker.DockerContainerLocation.getMappedPort(DockerContainerLocation.java:111) ~[brooklyn-clocker-docker-0.5.0-SNAPSHOT.jar:na]
    at brooklyn.location.docker.DockerContainerLocation.obtainPort(DockerContainerLocation.java:128) ~[brooklyn-clocker-docker-0.5.0-SNAPSHOT.jar:na]
    at brooklyn.event.basic.PortAttributeSensorAndConfigKey.convertConfigToSensor(PortAttributeSensorAndConfigKey.java:78) ~[brooklyn-core-0.7.0-SNAPSHOT.jar:na]
    at brooklyn.event.basic.PortAttributeSensorAndConfigKey.convertConfigToSensor(PortAttributeSensorAndConfigKey.java:1) ~[brooklyn-core-0.7.0-SNAPSHOT.jar:na]
    at brooklyn.event.basic.AttributeSensorAndConfigKey.getAsSensorValue(AttributeSensorAndConfigKey.java:93) ~[brooklyn-core-0.7.0-SNAPSHOT.jar:na]
    ... 12 common frames omitted
```

--- a/docker/src/main/java/brooklyn/entity/container/docker/DockerContainerImpl.java
+++ b/docker/src/main/java/brooklyn/entity/container/docker/DockerContainerImpl.java
@@ -332,9 +332,7 @@ public class DockerContainerImpl extends BasicStartableImpl implements DockerCon
         OsDetails osDetails = dockerHost.getDynamicLocation().getMachine().getMachineDetails().getOsDetails();
         if (osDetails.isMac()) {
             String address = dockerHost.execCommand("boot2docker ip");
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("The boot2docker IP address is {}", Strings.trim(address));
-            }
+            LOG.debug("The boot2docker IP address is {}", Strings.trim(address));
             return Strings.trim(address);
         } else {
             return null;
@@ -433,6 +431,9 @@ public class DockerContainerImpl extends BasicStartableImpl implements DockerCon
                 PortRange p = (PortRange) entity.getConfig(k);
                 if (p != null && !p.isEmpty()) ports.add(p.iterator().next());
             }
+        }
+        for (Entity child : entity.getChildren()) {
+            ports.addAll(getRequiredOpenPorts(child));
         }
         LOG.debug("getRequiredOpenPorts detected default {} for {}", ports, entity);
         return ports;

--- a/docker/src/main/java/brooklyn/location/docker/DockerContainerLocation.java
+++ b/docker/src/main/java/brooklyn/location/docker/DockerContainerLocation.java
@@ -52,6 +52,7 @@ import brooklyn.util.ssh.IptablesCommands.Chain;
 import brooklyn.util.ssh.IptablesCommands.Policy;
 import brooklyn.util.time.Duration;
 
+import com.google.common.base.Joiner;
 import com.google.common.base.Objects.ToStringHelper;
 import com.google.common.base.Predicates;
 import com.google.common.collect.ImmutableList;
@@ -104,9 +105,7 @@ public class DockerContainerLocation extends SshMachineLocation implements Suppo
     private void addIptablesRule(Integer port) {
         if (getOwner().getConfig(DockerInfrastructure.OPEN_IPTABLES)) {
             SshMachineLocation host = getOwner().getDockerHost().getDynamicLocation().getMachine();
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("Using iptables to add access for TCP/{} to {}", port, host);
-            }
+            LOG.debug("Using iptables to add access for TCP/{} to {}", port, host);
             List<String> commands = ImmutableList.of(
                     sudo("iptables -L INPUT -nv | grep -q 'tcp dpt:"+port+"'"),
                     format("if [ $? -eq 0 ]; then ( %s ); else ( %s ); fi",
@@ -125,8 +124,11 @@ public class DockerContainerLocation extends SshMachineLocation implements Suppo
         String containerId = getOwner().getContainerId();
         Map<Integer, Integer> mapping = JcloudsUtil.dockerPortMappingsFor(getOwner().getDockerHost().getJcloudsLocation(), containerId);
         Integer publicPort = mapping.get(portNumber);
-        if (LOG.isDebugEnabled()) {
-            LOG.debug("Docker mapped port {} to {} for Container {}", new Object[] { portNumber, publicPort, containerId });
+        if (publicPort == null) {
+            LOG.warn("Unable to map port {} for Container {}. Mappings: {}",
+                    new Object[]{portNumber, containerId, Joiner.on(", ").withKeyValueSeparator("=").join(mapping)});
+        } else {
+            LOG.debug("Docker mapped port {} to {} for Container {}", new Object[]{portNumber, publicPort, containerId});
         }
         return publicPort;
     }
@@ -226,9 +228,7 @@ public class DockerContainerLocation extends SshMachineLocation implements Suppo
 
     @Override
     public void close() throws IOException {
-        if (LOG.isDebugEnabled()) {
-            LOG.debug("Close called on Docker container {}: {}", machine, this);
-        }
+        LOG.debug("Close called on Docker container {}: {}", machine, this);
         try {
             machine.close();
             if (dockerContainer.getAttribute(DockerContainer.SERVICE_UP)) {

